src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -262,17 +262,16 @@ async Task<bool> Read(bool async, CancellationToken cancellationToken = default)
                     throw new ArgumentOutOfRangeException();
                 }
 
-                try
-                {
-                    var msg2 = await ReadMessage(async, CancellationToken.None);
-                    ProcessMessage(msg2);
-                    return msg2.Code == BackendMessageCode.DataRow;
-                }
-                catch
-                {
-                    State = ReaderState.Consumed;
-                    throw;
-                }
+                var msg2 = await ReadMessage(async, CancellationToken.None);
+                ProcessMessage(msg2);
+                return msg2.Code == BackendMessageCode.DataRow;
+            }
+            catch
+            {
+                State = ReaderState.Consumed;
+                // We cannot allow the reader to close the connection, otherwise there will be a concurency issues while disposing
+                await Cleanup(async, connectionClosing: true);
```

Comment:
I don't have all the context... But does this mean that if I specify CommandBehavior.CloseConnection but there is an error, the connection doesn't get closed? Or are we saying that the connection would be broken anyway if any of these errors occur?

Reply:
Yeah, it will not be closed - otherwise, there will be a problem with reader's disposing.

Reply:
But that's a problem and a behavioral change, no?

Reply:
It is a problem, yes.
Also, we're doing the same thing (moving close to dispose) in #3290

Reply:
So... should we think we think of another way to fix this issue, or are you saying we should merge this as-is and introduce this breaking change?

Reply:
It is not really a breaking change. Right now, the reader can be closed in 3 cases:
1) Call Reader.Close
2) Dispose reader
3) Error while reading

The only thing we're doing is removing the third one. Also, there is still the same bug in the first case.